[Core/Metrics] Accurately record used memory.

rkooo567 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

used_memory_ includes the fallback allocation, so we should subtract it here to calculate the exact available memory.

## Related issue number

Closes https://github.com/ray-project/ray/issues/24624

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

scv119 commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -736,7 +736,11 @@ std::string ObjectManager::DebugString() const {
+  // used_memory_ includes the fallback allocation, so we should subtract it here
+  // to calculate the exact available memory.
+  stats::ObjectStoreAvailableMemory().Record(
+      config_.object_store_memory - used_memory_ -
```
should it be `config_.object_store_memory - used_memory_ + plasma::plasma_store_runner->GetFallbackAllocated()` ?

clarkzinzow replied:
Yep it should be `config_.object_store_memory - used_memory_ - plasma::plasma_store_runner->GetFallbackAllocated()` or `config_.object_store_memory - (used_memory_ + plasma::plasma_store_runner->GetFallbackAllocated())`

clarkzinzow replied:
Yep it should be `config_.object_store_memory - used_memory_ - plasma::plasma_store_runner->GetFallbackAllocated()` or `config_.object_store_memory - (used_memory_ + plasma::plasma_store_runner->GetFallbackAllocated())`

rkooo567 commented:
Added tests. I also realize we need to handle `used_memory` properly too (if total - used is not avail, it is pretty weird), so I handled it and added a proper test

clarkzinzow reviewed (approved):
Nice, great test!
